software/contrib/gates_and_triggers.py

```diff
@@ -0,0 +1,161 @@
+
+    def on_rise(self):
+        """Handle the rising edge of the input signal
+        """
+        self.gate_on = True
+        self.on_incoming_rise_start_time = time.ticks_ms()
+        self.incoming_rise_out.voltage(OUTPUT_VOLTAGE)
+        self.gate_out.voltage(OUTPUT_VOLTAGE)
+
+        self.on_toggle()
+
+    def on_fall(self):
+        """Handle the falling edge of the input signal
+        """
+        self.on_incoming_fall_start_time = time.ticks_ms()
+        self.incoming_fall_out.voltage(OUTPUT_VOLTAGE)
+
+    def on_toggle(self):
+        """Handle toggling the toggle output
+        """
+        self.toggle_on = not self.toggle_on
+        if self.toggle_on:
+            self.toggle_out.voltage(OUTPUT_VOLTAGE)
+        else:
+            self.toggle_out.voltage(0)
+            self.toggle_fall_out.voltage(OUTPUT_VOLTAGE)
+            self.on_toggle_fall_start_time = time.ticks_ms()
+
+    def tick(self):
+        """Advance the clock and set the outputs high/low as needed
+        """
+        now = time.ticks_ms()
+
+        k1_percent = int(k1.percent() * 100)
+        k2_percent = int(k2.percent() * 100)
+
+        # if the user moved the knobs by more than 1% deactivate the screensaver
+        if abs(self.k1_percent - k1_percent) > 0 or abs(self.k2_percent - k2_percent) > 0:
+            ssoled.notify_user_interaction()
+
+        knob_lvl = k1_percent
+        cv_lvl = ain.percent() * (k2_percent * 0.02 - 1)
+        self.gate_duration = max(self.quadratic_knob(knob_lvl) + cv_lvl * 2000, TRIGGER_DURATION_MS)
+
+        if self.gate_on and time.ticks_diff(now, self.on_incoming_rise_start_time) > self.gate_duration:
+            self.gate_on = False
+            self.gate_out.voltage(0)
+            self.gate_fall_out.voltage(OUTPUT_VOLTAGE)
+            self.on_gate_fall_start_time = now
+
+        if time.ticks_diff(now, self.on_gate_fall_start_time) > TRIGGER_DURATION_MS:
+            self.gate_fall_out.voltage(0)
+
+        if time.ticks_diff(now, self.on_incoming_rise_start_time) > TRIGGER_DURATION_MS:
+            self.incoming_rise_out.voltage(0)
+
+        if time.ticks_diff(now, self.on_incoming_fall_start_time) > TRIGGER_DURATION_MS:
+            self.incoming_fall_out.voltage(0)
+
+        if time.ticks_diff(now, self.on_toggle_fall_start_time) > TRIGGER_DURATION_MS:
+            self.toggle_fall_out.voltage(0)
+
+        self.k1_percent = k1_percent
+        self.k2_percent = k2_percent
+
+    def quadratic_knob(self, x):
+        """Some magic math to give us a quadratic response on the knob percentage
+
+        This gives us 10ms @ 0% and 2000ms @ 100% with greater precision at the higher end
+        where the differences will be more noticeable
+        """
+        if x <= 0:
+            return 10
+        return 199 * sqrt(x) + 10
+
+    def main(self):
+        while(True):
+            self.tick()
+
+            ssoled.fill(0)
+            ssoled.centre_text(f"Gate: {self.gate_duration:0.0f}ms")
```

Comment:
[comment] Might be the simplest screen UI in the entire repo!

Reply:
I was honestly debating even bothering to have a GUI! The original version didn't, but then I had to clear the splash screen and it seemed easier to just have _something_ on screen